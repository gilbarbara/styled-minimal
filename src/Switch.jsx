import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { getColor, px, themeGet } from './utils/helpers';
import { sizesAllPropTypes, variantPropTypes } from './utils/system';

import Box from './Box';

const styles = (props: Object): string => {
  const { size } = props;
  const sizes = themeGet(props, 'switchSizes');

  return css`
    cursor: pointer;
    height: ${px(sizes[size].height)};
    position: relative;
    user-select: none;
    vertical-align: middle;
    width: ${px(sizes[size].width)};
  `;
};

const stylesTrack = (props: Object): string => {
  const { size, status } = props;
  const sizes = themeGet(props, 'switchSizes');
  const themeColor = getColor(props);

  return css`
    background-color: ${status ? themeColor : '#ccc'};
    border-radius: ${px(sizes[size].borderRadius)};
    bottom: 0;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
  `;
};

const stylesButton = (props: Object): string => {
  const { size, status } = props;
  const sizes = themeGet(props, 'switchSizes');

  return css`
    background-color: #fff;
    border-radius: 50%;
    bottom: ${px(sizes[size].space)};
    left: ${status ? '50%' : px(sizes[size].space)};
    position: absolute;
    top: ${px(sizes[size].space)};
    transition: left 0.1s ease;
    width: ${px(sizes[size].height - sizes[size].space * 2)};
  `;
};

const StyledInput = styled.input`
  bottom: 0;
  left: 0;
  opacity: 0;
  position: absolute;
  right: 0;
  top: 0;
`;

const StyledTrack = styled.span`
  ${stylesTrack};
`;

const StyledButton = styled.span`
  ${stylesButton};
`;

export const StyledSwitch = styled(Box)`
  ${styles};
`;

class Switch extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      status: props.value,
    };
  }

  static propTypes = {
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    size: sizesAllPropTypes,
    value: PropTypes.bool,
    variant: variantPropTypes,
    ...Box.propTypes,
  };

  static defaultProps = {
    onChange: () => {},
    size: 'md',
    value: false,
    variant: 'primary',
  };

  componentDidUpdate(prevProps) {
    const { value } = this.props;

    if (prevProps.value !== value) {
      this.update(value);
    }
  }

  update = value => {
    this.setState({ status: value });
  };

  handleClick = () => {
    const { status } = this.state;
    const { onChange } = this.props;

    this.setState({
      status: !status,
    });

    onChange(!status);
  };

  render() {
    const { name, ...rest } = this.props;
    const { status } = this.state;

    return (
      <StyledSwitch status={status} {...rest} onClick={this.handleClick}>
        <StyledInput type="hidden" name={name} value={status} />
        <StyledTrack status={status} {...rest} />
        <StyledButton status={status} {...rest} />
      </StyledSwitch>
    );
  }
}

export default Switch;
