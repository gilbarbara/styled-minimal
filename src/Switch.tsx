import * as React from 'react';
import styled, { css } from 'styled-components';

import { getColor, getTheme, isBoolean, px } from './utils/helpers';
import { baseStyles, sx, system } from './utils/system';
import { Label } from './Label';

import { SwitchProps } from './types';

export interface SwitchPropsWithRef extends SwitchProps {
  innerRef?: React.Ref<HTMLInputElement>;
}

interface InnerProps extends Omit<SwitchProps, 'name'> {
  status: boolean;
}

const StyledInput = styled('input')`
  bottom: 0;
  left: 0;
  opacity: 0;
  position: absolute;
  right: 0;
  top: 0;
`;

const StyledTrack = styled('span')<InnerProps>(props => {
  const { size, status } = props;
  const sizes = getTheme(props, 'switchSizes');
  const color = status ? getColor(props) : '#ccc';

  return css`
    background-color: ${color};
    border-radius: ${px(sizes[size!].borderRadius)};
    bottom: 0;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
  `;
});

const StyledButton = styled('span')<InnerProps>(props => {
  const { disabled, size, status } = props;
  const sizes = getTheme(props, 'switchSizes');
  const currentSize = sizes[size!];

  return css`
    background-color: #fff;
    border-radius: 50%;
    bottom: ${px(currentSize.space)};
    left: ${status ? '50%' : px(currentSize.space)};
    opacity: ${disabled ? 0.7 : 1};
    position: absolute;
    top: ${px(currentSize.space)};
    transition: left 0.1s ease;
    width: ${px(currentSize.height - currentSize.space * 2)};
  `;
});

export const StyledSwitch = styled('div')<InnerProps>(
  props => {
    const { children, disabled, size } = props;
    const sizes = getTheme(props, 'switchSizes');
    const currentSize = sizes[size!];

    return css`
      ${baseStyles};
      cursor: ${disabled ? 'default' : 'pointer'};
      height: ${px(currentSize.height)};
      margin-right: ${children ? '8px' : 0};
      opacity: ${disabled ? 0.8 : 1};
      position: relative;
      user-select: none;
      vertical-align: middle;
      width: ${px(currentSize.width)};
    `;
  },
  sx,
  system,
);

class SwitchClass extends React.PureComponent<SwitchPropsWithRef, { status: boolean }> {
  constructor(props: SwitchProps) {
    super(props);

    this.state = {
      status: isBoolean(props.checked) ? props.checked : props.defaultChecked!,
    };
  }

  static defaultProps = {
    defaultChecked: false,
    disabled: false,
    size: 'md',
    variant: 'primary',
  };

  componentDidUpdate(prevProps: SwitchProps) {
    const { checked } = this.props;

    if (isBoolean(checked) && prevProps.checked !== checked) {
      this.update(checked);
    }
  }

  update = (status: boolean) => {
    this.setState({ status });
  };

  handleClick = () => {
    const { status } = this.state;
    const { checked, disabled, onChange } = this.props;

    if (isBoolean(checked) || disabled) {
      return;
    }

    this.update(!status);

    /* istanbul ignore else */
    if (onChange) {
      onChange(status ? 'off' : 'on');
    }
  };

  render() {
    const {
      checked,
      disabled,
      children,
      innerRef,
      name,
      onChange,
      size,
      variant,
      ...rest
    } = this.props;
    const { status } = this.state;

    const component = (
      <StyledSwitch
        disabled={disabled}
        size={size}
        status={status}
        role="switch"
        aria-checked={status}
        {...rest}
        onClick={!children ? this.handleClick : undefined}
      >
        <StyledInput
          disabled={disabled}
          name={name}
          ref={innerRef}
          type="hidden"
          value={status ? 'on' : 'off'}
        />
        <StyledTrack size={size} status={status} variant={variant} />
        <StyledButton disabled={disabled} size={size} status={status} variant={variant} />
      </StyledSwitch>
    );

    if (children) {
      return (
        <Label
          alignItems="center"
          cursor={disabled ? 'default' : 'pointer'}
          onClick={this.handleClick}
        >
          {component}
          {children}
        </Label>
      );
    }

    return component;
  }
}

export const Switch = React.forwardRef<HTMLInputElement, SwitchPropsWithRef>((props, ref) => (
  <SwitchClass innerRef={ref} {...props} />
));
