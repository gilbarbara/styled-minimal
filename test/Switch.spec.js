import React from 'react';
import { mount } from 'enzyme';

import Switch from '../src/Switch';

describe('Switch', () => {
  const wrapper = mount(<Switch name="toggle" value={true} variant="indigo" />);

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
