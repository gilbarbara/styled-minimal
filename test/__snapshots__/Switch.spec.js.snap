// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render properly 1`] = `
.c1 {
  bottom: 0;
  left: 0;
  opacity: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c2 {
  background-color: #ccc;
  border-radius: 14px;
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c3 {
  background-color: #fff;
  border-radius: 50%;
  bottom: 3px;
  left: 3px;
  position: absolute;
  top: 3px;
  -webkit-transition: left 0.1s ease;
  transition: left 0.1s ease;
  width: 20px;
}

.c0 {
  box-sizing: border-box;
  cursor: pointer;
  height: 26px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  width: 47px;
}

<StyledSwitch
  onChange={[Function]}
  onClick={[Function]}
  size="md"
  status={true}
  value={true}
  variant="indigo"
>
  <div
    className="c0"
    onChange={[Function]}
    onClick={[Function]}
    size="md"
    value={true}
  >
    <StyledInput
      name="toggle"
      type="hidden"
      value={true}
    >
      <input
        className="c1"
        name="toggle"
        type="hidden"
        value={true}
      />
    </StyledInput>
    <StyledTrack
      onChange={[Function]}
      size="md"
      value={true}
      variant="indigo"
    >
      <span
        className="c2"
        onChange={[Function]}
        size="md"
        value={true}
      />
    </StyledTrack>
    <StyledButton
      onChange={[Function]}
      size="md"
      value={true}
      variant="indigo"
    >
      <span
        className="c3"
        onChange={[Function]}
        size="md"
        value={true}
      />
    </StyledButton>
  </div>
</StyledSwitch>
`;
